File: src/leetCode/easy_20_validParentheses/validParentheses.go
package validparentheses

// https://leetcode-cn.com/problems/valid-parentheses/
// https://github.com/azl397985856/leetcode/blob/master/problems/20.valid-parentheses.md

func isValid(s string) bool {
	n := len(s)

	if n == 0 {
		return true
	}

	if n%2 != 0 {
		return false
	}

	stack := []byte{}

	pairs := map[byte]byte{
		')': '(',
		'}': '{',
		']': '[',
	}

	for i := 0; i < n; i++ {
		expectedLeft, isRight := pairs[s[i]]

		if isRight {
			if len(stack) == 0 || stack[len(stack)-1] != expectedLeft {
				return false
			}

			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, s[i])
		}
	}

	return len(stack) == 0
}

func isValidA(s string) bool {
	n := len(s)

	if n == 0 {
		return true
	}

	if n%2 != 0 {
		return false
	}

	pairs := map[byte]byte{
		'(': ')',
		'[': ']',
		'{': '}',
	}

	lps := make([]byte, 0, n/2)
	for i := 0; i < n; i++ {
		p := s[i]

		if _, ok := pairs[p]; ok {
			lps = append(lps, p)
		} else {
			if len(lps) == 0 || pairs[lps[len(lps)-1]] != p {
				return false
			}

			lps = lps[:len(lps)-1]
		}
	}

	return len(lps) == 0
}

File: src/leetCode/easy_20_validParentheses/validParentheses.ts

/**
 * @param {string} s
 * @returns {boolean}
 */
function isValid(s: string): boolean {
  if (s === '') {
    return true;
  }

  if (s.length % 2 !== 0) {
    return false;
  }

  const pairs = new Map<string, string>([
    ['(', ')'],
    ['[', ']'],
    ['{', '}'],
  ]);

  const stack: string[] = [];

  [...s].forEach((p) => {
    if (pairs.has(p)) {
      stack.push(p);
    } else {
      if (stack.length === 0) {
        return false
      }

      const last: string = stack[stack.length - 1];

      if (p !== pairs.get(last)) {
        return false;
      }

      stack.pop();
    }
  });

  return stack.length === 0;
};
